### PR TITLE
Fixed bugs with different enum formats, closes #3

### LINKED ISSUE #3: Internally tagged enums are not working
I was trying this library out and I noticed that internal tagging is not deserializing properly. I get runtime errors after simply serializing to a JsValue and deserializing from it.

I looked through your code and found the internally tagged enum test, but it looks like you guys aren't expanding the `#[serde(tag = "tag")]` attribute in your `test_enum! { ... }` macro.

I started getting compile errors in the tests/serde.rs file after I added the expansion to the macro_rules block:

```rust
macro_rules! test_enum {
    ($(# $attr:tt)* $name:ident) => {{
        #[derive(Debug, PartialEq, Serialize, Deserialize)]
        $(# $attr)*    // RIGHT HERE
        enum $name<A, B> {
            Unit,
            Newtype(A),
            Tuple(A, B),
            Struct { a: A, b: B },
        }

        test_via_json($name::Unit::<(), ()>);
        test_via_json($name::Newtype::<_, ()>("newtype content".to_string()));
        test_via_json($name::Tuple("tuple content".to_string(), 42));
        test_via_json($name::Struct {
            a: "struct content".to_string(),
            b: 42,
        });
    }};
}
```

Console output:

```
serde-wasm-bindgen$ wasm-pack test --node
[INFO]: Checking for the Wasm target...
   Compiling serde-wasm-bindgen v0.1.3 (serde-wasm-bindgen)
error: #[serde(tag = "...")] cannot be used with tuple variants

error[E0277]: the trait bound `enums::InternallyTagged<(), ()>: enums::_IMPL_DESERIALIZE_FOR_ExternallyTagged::_serde::Serialize` is not satisfied
   --> tests/serde.rs:89:23
    |
39  |   fn test_via_json<T>(value: T)
    |      -------------
40  |   where
41  |       T: Serialize + DeserializeOwned + PartialEq + Debug,
    |          --------- required by this bound in `test_via_json`
...
89  |           test_via_json($name::Unit::<(), ()>);
    |                         ^^^^^^^^^^^^^^^^^^^^^ the trait `enums::_IMPL_DESERIALIZE_FOR_ExternallyTagged::_serde::Serialize` is not implemented for `enums::InternallyTagged<(), ()>`
...
218 | /     test_enum! {
219 | |         #[serde(tag = "tag")]
220 | |         InternallyTagged
221 | |     }
    | |_____- in this macro invocation
```

These errors do make sense because internally tagged enums don't work with tuple variants. It's just that they weren't being tested before.

I'm looking at this bug to see if I can fix it, but unfortunately I don't know much about serde. I'm not sure if it's a simple fix or not, but if I figure anything out I'll open a PR.

In the meantime, if you guys have any thoughts or guidance on this, I'd appreciate it!

### PR DESCRIPTION
**Pull request by awestlake87**
It looks like the problem mainly came from the deserialize_any function. Apparently, serde does some intermediate buffering of the fields in internally-tagged and untagged enums before completely deserializing them. Because of this, the deserialize_any field needs to be able to differentiate between integers, floats, arrays, and maps.

I changed this function to call the visit_i64 function if the JsValue is a safe integer. The serde ContentDeserializer would not cast a f64 into an i64 during deserialization, so I prioritized the i64 if it meets the criteria of a [safe integer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger). Serde will later cast this back into a float if necessary.

Additionally, in order to get the content of those enum variants working, I had to add support for maps and sequences in deserialize_any.

In the tests, I removed the variants that were incompatible with internally-tagged enums and added special cases in assert_json for untagged enum variants. These variants are not necessarily objects, so they did not play well with JSON::stringify (at least for me).

### CODE REVIEW

File: tests/serde.rs [hunk 1]
```diff
@@ -214,10 +230,24 @@ fn enums() {
...
-        InternallyTagged
+
+    #[derive(Debug, PartialEq, Serialize, Deserialize)]
+    #[serde(tag = "tag")]
+    enum InternallyTagged<A, B> {
+        Unit,
+        Struct { a: A, b: B },
     }
+
+    test_via_json(InternallyTagged::Unit::<(), ()>);
```
[the remark this answers is not in the document]
Comment (awestlake87) [reply tests/serde.rs]: It can't use the `test_enum!` macro as it is for internally tagged enums since serde doesn't support internal tags on tuple variants. 

There was also a problem with the NewType variant since internally tagged enums work with a struct as the new type, but not something like a string (which is what the `test_enum!` macro was testing it with). This is because serde has to insert the tag inside the key/value pairs of the struct in a new-type variant.

I'm open to testing this enum in a different way though.

File: tests/serde.rs [hunk 2]
```diff
@@ -28,10 +28,25 @@ fn assert_json<R>(lhs_value: JsValue, rhs: R)
...
-    assert_eq!(
-        js_sys::JSON::stringify(&lhs_value).unwrap(),
-        serde_json::to_string(&rhs).unwrap(),
-    );
+    if lhs_value.is_object() {
+        assert_eq!(
+            js_sys::JSON::stringify(&lhs_value).unwrap(),
+            serde_json::to_string(&rhs).unwrap(),
+        );
+    } else if lhs_value.is_string() {
```
[the remark this answers is not in the document]
Comment (awestlake87) [reply tests/serde.rs]: It's been awhile since I made this change. I'll try reverting it and get back to you on whether it was really necessary.

File: tests/serde.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (RReverser) [reply tests/serde.rs]: Ah yeah, this makes sense. Thanks for clarification.

File: src/de.rs [hunk 3]
```diff
@@ -236,9 +236,17 @@ impl<'de> de::Deserializer<'de> for Deserializer {
         } else if let Some(v) = self.value.as_bool() {
             visitor.visit_bool(v)
         } else if let Some(v) = self.value.as_f64() {
-            visitor.visit_f64(v)
+            if js_sys::Number::is_safe_integer(&self.value) {
```
Comment (RReverser) [inline src/de.rs]: Are these conversions covered by new tests somehow?

Generally I'd be wary of extending `deserialize_any` with casts, and prefer that users specify explicit destination types, unless there is a compelling reason to use dynamic heuristics.

Comment (awestlake87) [reply src/de.rs]: Yeah, this gets into the weeds of how serde handles internally tagged enums. Basically what happens is serde buffers the variant into an intermediate Content structure. When it encounters a JS number, it stores it as an f64, which ends up breaking deserialization for integer types when it converts this intermediate structure to the one we actually want to deserialize.

The same is true for sequences and objects, which is why I added checks for the Array and Object types.

In the test cases, I test one with type B as a float, and one with type B as an int. I believe this also covers the Object type, since the struct has to be deserialized as a map first, but I'm not certain. We will probably want to add a test for a sequence type (I guess I only verified that this worked in my project).

Comment (RReverser) [reply src/de.rs]: Ughh this is unfortunate, but makes sense...

File: tests/serde.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (awestlake87) [reply tests/serde.rs]: Sorry it's been awhile since I responded. I've had a busy month so far.

Anyway, I looked into it, and it looks like the `lhs_value.is_string()` branch can actually be combined with the `lhs_value.is_object()` branch. However, the `lhs_value.is_undefined() || lhs_value.is_null()` is actually necessary since `js_sys::JSON::stringify(&lhs_value).unwrap()` panics when `lhs_value` is null.

I'll combine the redundant branches and make another commit

[the remark this answers is not in the document]
Comment (awestlake87) [reply tests/serde.rs]: Here it is: c41d8e0

[the remark this answers is not in the document]
Comment (RReverser) [reply tests/serde.rs]: > is actually necessary since js_sys::JSON::stringify(&lhs_value).unwrap() panics when lhs_value is null

Ah - the problem is that `JSON.stringify(undefined)` in JavaScript returns `undefined` and not a string. However, `JSON.stringify(null)` should work fine.

[the remark this answers is not in the document]
Comment (awestlake87) [reply tests/serde.rs]: Ok, I simplified the condition to only include a separate case for `undefined` values: c326b7f.

File: src/de.rs (hunk 3 above)
Comment (RReverser) [reply src/de.rs]: Coming back to this - can you please make sure to cover these branches by tests (outside of internally tagged enums)? I want to make sure that assumptions about representation of new types are clearly defined for future regressions. Thanks!

Comment (awestlake87) [reply src/de.rs]: I'm running into a few problems with this. I added two variants to the `test_enum!` macro, making it look like this:

```rust
macro_rules! test_enum {
    ($(# $attr:tt)* $name:ident) => {{
        #[derive(Debug, PartialEq, Serialize, Deserialize)]
        $(# $attr)*
        enum $name<A, B> where A: Debug + Hash + Eq {
            Unit,
            Newtype(A),
            Tuple(A, B),
            Struct { a: A, b: B },
            Sequence(Vec<A>),
            Map(HashMap<A, B>),
        }

        test_via_json($name::Unit::<String, i32>);
        test_via_json($name::Newtype::<_, i32>("newtype content".to_string()));
        test_via_json($name::Tuple("tuple content".to_string(), 42));
        test_via_json($name::Struct {
            a: "struct content".to_string(),
            b: 42,
        });
        test_via_json($name::<i32, i32>::Sequence(vec![52, 1, -124, 23, -65]));
        test_via_json($name::<String, i32>::Map(
            vec![("a".to_string(), 12), ("abc".to_string(), -1161), ("b".to_string(), 64)].into_iter().collect()
        ));
    }};
}
```

This should at least touch on some sequences and maps in non-internally tagged enums. However, I ran into the following error with this setup:

```
failures:

---- serde::enums output ----
    error output:
        panicked at 'assertion failed: `(left == right)`
          left: `"{\"Map\":{}}"`,
         right: `"{\"Map\":{\"abc\":-1161,\"b\":64,\"a\":12}}"`', tests/serde.rs:37:9
        
        Stack:
        
        Error
            at module.exports.__wbg_new_59cb74e423758ede (serde-wasm-bindgen/target/wasm32-unknown-unknown/wbg-tmp/wasm-bindgen-test.js:524:21)
            at console_error_panic_hook::Error::new::h274f015c600dfb3e (wasm-function[6169]:0x2018c8)
            at console_error_panic_hook::hook_impl::h3ca87fb44cb9eea6 (wasm-function[1239]:0x151671)
            at console_error_panic_hook::hook::h2b7999185aaf7486 (wasm-function[6797]:0x20b038)
            at core::ops::function::Fn::call::hc246242b419002f9 (wasm-function[6193]:0x201f6b)
            at std::panicking::rust_panic_with_hook::h8540df5fb5cc4da7 (wasm-function[1847]:0x181b27)
            at std::panicking::continue_panic_fmt::h3082507e3bf1767e (wasm-function[3830]:0x1cf7a9)
            at std::panicking::begin_panic_fmt::hd9925b25492d4f84 (wasm-function[6026]:0x1ff2bf)
            at serde::assert_json::h2af830e0b2abbb33 (wasm-function[186]:0x6ba78)
            at serde::test_via_json::hd30427c9dd2fd6e6 (wasm-function[1698]:0x1777a1)
        
        
    
    JS exception that was thrown:
        RuntimeError: unreachable
            at __rust_start_panic (wasm-function[7985]:0x2160dc)
            at rust_panic (wasm-function[7342]:0x211ef4)
            at std::panicking::rust_panic_with_hook::h8540df5fb5cc4da7 (wasm-function[1847]:0x181b4e)
            at std::panicking::continue_panic_fmt::h3082507e3bf1767e (wasm-function[3830]:0x1cf7a9)
            at std::panicking::begin_panic_fmt::hd9925b25492d4f84 (wasm-function[6026]:0x1ff2bf)
            at serde::assert_json::h2af830e0b2abbb33 (wasm-function[186]:0x6ba78)
            at serde::test_via_json::hd30427c9dd2fd6e6 (wasm-function[1698]:0x1777a1)
            at serde::enums::h8c39b7b5568167b3 (wasm-function[89]:0x1eb44)
            at core::ops::function::FnOnce::call_once::hf429f4f99c997298 (wasm-function[7207]:0x21078e)
            at wasm_bindgen_test::__rt::Context::execute_sync::{{closure}}::hc76b1a4e656b2ee3 (wasm-function[5241]:0x1f0cde)

failures:

    serde::enums

test result: FAILED. 10 passed; 1 failed; 0 ignored
```

Not sure why the `HashMap` is being skipped over here, but it appears to be another bug (unless I'm missing something). Note that serializing structs works perfectly fine, but not `HashMap`, which I believe works fine for `serde_json`.

Another thing I saw was due to the way `assert_json` checks the serialization results. When I add another test case to `test_enum!` for sequences of floats:
```rust
test_via_json(InternallyTagged::<f32, ()>::Sequence(vec![12.4, 11.0, -124.5, 232.2, -65.56]));
```
 I get this error due to floating points not being serialized exactly the same by `serde_json` and `js_sys::JSON`:
```
---- serde::enums output ----
    error output:
        panicked at 'assertion failed: `(left == right)`
          left: `"{\"Sequence\":[12.399999618530273,11,-124.5,232.1999969482422,-65.55999755859375]}"`,
         right: `"{\"Sequence\":[12.4,11.0,-124.5,232.2,-65.56]}"`', tests/serde.rs:37:9
...
```
This indicates that we need a more robust method of asserting that values are serialized and deserialized properly in order to provide more comprehensive coverage.

Comment (RReverser) [reply src/de.rs]: > Note that serializing structs works perfectly fine, but not HashMap, which I believe works fine for serde_json.

I suspect the culprit is different JavaScript representations for HashMap and structs, which now conflict as part of `Content` / `any` deserialization. Try printing intermediate representations of each value after it's serialised?

Comment (RReverser) [reply src/de.rs]: By the way, more generally what you've described sounds like yet another usecase of https://github.com/serde-rs/serde/issues/1183.

Comment (RReverser) [reply src/de.rs]: Hi! Did you ever happen to find a fix to this issue?

File: tests/serde.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (RReverser) [reply tests/serde.rs]: Sorry I missed this comment before. This update looks good!

File: src/de.rs (hunk 3 above)
Comment (awestlake87) [reply src/de.rs]: I haven't yet, but I've been meaning to come back to this. I'll try to take a look at it sometime this weekend.

Comment (awestlake87) [reply src/de.rs]: So, the problem seems to be `js_sys::Map`. Apparently JSON::stringify doesn't work with Map since it's possible to have non-string keys in a Map, which is unsupported by JSON.

I don't think this is a deal-breaker for `js_sys::Map`, it just means we need to use something other than JSON to verify it ([Relevant stack overflow question](https://stackoverflow.com/questions/29085197/how-do-you-json-stringify-an-es6-map)). We need something similar for other test cases such as the sequence test that I mentioned above.
 
I haven't tested serialization/deserialization for the `js_sys::Map` to see if it works. I didn't see any red flags from a sight-read of the `ser.rs`, though.

Comment (RReverser) [reply src/de.rs]: Ah yeah, that makes sense. For `js_sys::Map` we might just convert it to an array using `Array::from` before stringifying.

Comment (awestlake87) [reply src/de.rs]: Ok I added some more tests. I think the string serialization is a bit fragile, so I went ahead and wrote some functions to check for structural equality of JS values. These should be more robust with rounding errors and should open the door for testing binary data as well. I think it can act as a full replacement for the `test_via_json` function if that's the route you guys want to take.

Looks like everything is working now, though.

Comment (RReverser) [reply src/de.rs]: The string tests were serving not only to test structural equality, but also to ensure backwards compatibility with serde_json for the common subset of supported types.

This is because wasm-bindgen currently uses serde-json under the hood, and we want to ensure a smooth transition between two projects.

Comment (RReverser) [reply src/de.rs]: It feels like this PR does a great thing but keeps growing to include unrelated changes (like general tests overhaul), making it harder to just merge quickly :(

If you could please separate changes and keep this one simple and cover only the actual issue, that would be highly appreciated and make reviewing and merging much faster.

Comment (RReverser) [reply src/de.rs]: Or, I guess, I might try to cherry-pick the relevant changes myself if I get to it earlier...

Comment (awestlake87) [reply src/de.rs]: Is it just the structural equality or were you wanting to separate the changes for sequences and maps too?

Comment (RReverser) [reply src/de.rs]: Sequences and maps are fine, but I'd keep the relevant test changes in this PR. The other changes could be sent and discussed in a separate PR, but for now just want to make progress on fixing the original issue :)

Comment (awestlake87) [reply src/de.rs:239]: Alright, I found some time to remove the structural equality testing stuff. 

Converting Maps to Objects with `Object::from_entries` proved to be a bit trickier than anticipated since externally/adjacently tagged enums use the Map a layer deeper than internal/untagged enums. To fix this, I recurse through the object and replace all instances of Map with an Object. 

I also changed the f32 sequences to i32 to avoid rounding differences in the stringified versions. There's not a whole lot I can do about that since `JSON.stringify` and `serde_json::to_string` just have fundamentally different behavior.

If you look at the overall diffs for the changesets, I think it's about as minimal as I can get it. Let me know if there's anything else we need to do before merging.

Comment (RReverser) [reply src/de.rs:239]: Thanks, I've pushed a commit to exclude any iterables from `deserialize_any` to limit the scope but otherwise looks much simpler.

Can you please remind why do we still need `recurse_and_replace_maps` in tests? Is it something that end users would need to do as well? If so, similar logic probably has to be moved to the library code instead.

Comment (awestlake87) [reply src/de.rs:239]: I don't think it's something that end users will need. The reason why I had to do that in the tests is because `JSON.stringify` doesn't serialize the contents of `Map` objects. So the following test panics:

```rust
#[derive(Serialize, Deserialize)]
enum Value {
    Map(HashMap<String, i32>)
}

#[test]
fn test() {
    let value = Value::Map(vec![("a".to_string(), 4)].into_iter().collect());
    assert_eq!(
        js_sys::JSON::stringify(&serde_wasm_bindgen::to_value(&value).unwrap()).unwrap(),
        serde_json::to_string(&value).unwrap()
    );
}
```

With the error:

```
    error output:
        panicked at 'assertion failed: `(left == right)`
          left: `"{\"Map\":{}}"`,
         right: `"{\"Map\":{\"a\":4}}"`'
```

Because of this, we discussed converting `js_sys::Map` to an array using `Array::from` or an object using `Object::from_entries` before stringifying.

However, this posed a problem because sometimes the map would be at different levels in the object tree. For example, if we changed the above enum to an untagged enum, it would give this error instead:

```
    error output:
        panicked at 'assertion failed: `(left == right)`
          left: `"{}"`,
         right: `"{\"a\":4}"`'
```

Or if the enum contained a `HashMap` or a struct with a `HashMap` deeper within the enum variant, the `Map` would be even deeper in the object tree. For this reason, I couldn't just replace the Map at the top level or at predefined locations in the JsValue, so I wrote that `recurse_and_replace_maps` function to handle this conversion for `js_sys::Map` for *all* maps in the object tree.

Comment (RReverser) [reply src/de.rs:239]: So I didn't immediately realise that internal tags for maps mean that it's actually producing JavaScript `Map`s with one of the items being a tag... While in principle this works, it feels super awkward to commit to this and I wonder if there are any real-world use-cases that want this.

I'd imagine that most use-cases that do want internally tagged enums with arbitrary fields would rather use JS object representation (with string-only keys) instead of an ES6 Map, but that would be a separate feature - perhaps a namespace that could be used with `serde(with = ...)` attribute to ser-/deserialize maps as JS objects.

If you don't know of cases that do need ES6 `Map` with internally tagged enums, maybe let's remove that from PR for now?

(Btw, I'm sorry for my slow responses here - since I switched jobs, it's a bit harder to find spare time for some OSS projects.)

**Comment by awestlake87** [comment]
I personally dont know of any concrete use cases for it. Maybe it can be helpful if you know the tag of a structure but the other fields have names only known at runtime?


In any case, i believe that this is just doing what serde_json is doing. If the goal is parity with the serde_json, then i think this has to work.

File: src/de.rs (hunk 3 above)
Comment (awestlake87) [reply src/de.rs:239]: No worries, I haven't been the best about getting back to this either.

I personally dont know of any concrete use cases for it. Maybe it can be helpful if you know the tag of a structure but the other fields have names only known at runtime?

In any case, i believe that this is just doing what serde_json is doing. If the goal is parity with the serde_json, then i think this has to work.

Comment (RReverser) [reply src/de.rs:239]: The difference is, serde_json doesn't support native JS types like Map / Set, while we do, so for these types we can already deviate.

For compatibility with serde_json we would probably need a separate helper to serialize maps with string keys as objects (like one I mentioned above), but I think we can do that outside of this PR.

Comment (RReverser) [reply src/de.rs:239]: > We could change maps to be an Object by default, but that would likely screw someone up downstream. Do you have anything in mind?

Looks like we were writing past each other 😅 Yeah, I don't think we can do change like that by default (even if we could technically, which is also complicated). Best to leave this as an alternative format with explicit opt-in.

Comment (awestlake87) [reply src/de.rs:239]: Yeah i agree. 

Im not super familiar with serde, though. Maybe there's some weird way we can show a compiler error if internal tags are paired with a map on a new-type variant?

Comment (RReverser) [reply src/de.rs:239]: I don't think so :( As you discovered in this PR already, Serde does some implicit conversions when internal tags are involved with no way to hook into them at the moment.

By https://github.com/cloudflare/serde-wasm-bindgen/pull/4/commits/df09c8b67bee7eb6e24c5d915b5c4f4f22a64307 I've already semi-accidentally disabled internal tags + Maps during deserialization, but I don't know of way to detect that during serialization as well.

I'd be happy to be proven wrong if such way exists, but if not, I think we can at least remove them from tests and not guarantee such conversion in docs.

Comment (awestlake87) [reply src/de.rs:239]: Ok i can take a look at removing that from the tests later this evening. Im not sure which places to update in the docs though

Comment (RReverser) [reply src/de.rs:239]: Just the [Supported types in README](https://github.com/cloudflare/serde-wasm-bindgen#supported-types).

Comment (awestlake87) [reply src/de.rs:239]: So after looking into df09c8b, I'm a little unclear as to why iterables get excluded there. I agree that we should have a serialization switch for `Map` vs `Object`, but if we don't at the moment, then isn't it ok just to allow deserialization from iterables?

It might not be the best policy, but it wouldn't be unreasonable until a `serde(with = ...)` switch is implemented or https://github.com/serde-rs/serde/issues/1183 gets resolved in my opinion. At the very least, it might save someone a gotcha if they just naively replace their `serde_json` code with `serde_wasm_bindgen`.

I suspended that changeset and ran this test and it seemed to do fine:
```rust
assert_eq!(
    InternallyTagged::Map(
        vec![
            ("a".to_string(), 12), 
            ("abc".to_string(), -1161), 
            ("b".to_string(), 64)
        ].into_iter().collect()
    ), 
    from_value::<InternallyTagged<String, i32>>(
        js_sys::eval(r"
            var m = new Map(); 
            m.set('tag', 'Map'); 
            m.set('a', 12);
            m.set('abc', -1161);
            m.set('b', 64);
            m
        ").unwrap()
    ).unwrap()
);
```

I'll upload the deserialization test and the README change so you can look at it, though.

Comment (RReverser) [reply src/de.rs:239]: My hesitation with allowing such conversions that is that someone will most likely start relying on that behaviour, and removing it afterwards in a favour of "the right way" will be a breaking change.

I'd rather be as restrictive as possible and throw an explicit error, until someone comes up with a use-case where this is important, rather than silently allow some conversions through and deal with the challenge of deprecating them later.

Comment (awestlake87) [reply src/de.rs:239]: Another thing to consider is that excluding Iterables also affects untagged enums since they go through `deserialize_any` as well. So this test fails with df09c8b as well:

```rust
    #[derive(Debug, PartialEq, Serialize, Deserialize)]
    #[serde(untagged)]
    enum Untagged<A, B>
    where
        A: Ord,
    {
        Unit,
        Sequence { seq: Vec<A> },
        Map(BTreeMap<A, B>),
    }
    assert_eq!(
        Untagged::Map(
            vec![
                ("a".to_string(), 12), 
                ("abc".to_string(), -1161), 
                ("b".to_string(), 64)
            ].into_iter().collect()
        ), 
        from_value::<Untagged<String, i32>>(
            js_sys::eval(r"
                var m = new Map(); 
                m.set('tag', 'Map'); 
                m.set('a', 12);
                m.set('abc', -1161);
                m.set('b', 64);
                m
            ").unwrap()
        ).unwrap()
    );
```

Comment (awestlake87) [reply src/de.rs:239]: Oof and basically maps at any level in the variant. This also fails:

```rust
    #[derive(Debug, PartialEq, Serialize, Deserialize)]
    #[serde(tag = "tag")]
    enum InternallyTagged<A, B>
    where
        A: Ord,
    {
        Unit,
        Struct { a: A, b: B },
        Sequence { seq: Vec<A> },
        Map(BTreeMap<A, B>),
        DeeperMap {
            map: BTreeMap<A, B>,
        }
    }

    assert_eq!(
        InternallyTagged::DeeperMap {
            map: vec![
                ("a".to_string(), 12), 
                ("abc".to_string(), -1161), 
                ("b".to_string(), 64)
            ].into_iter().collect()
        }, 
        from_value::<InternallyTagged<String, i32>>(
            js_sys::eval(r"
                var m = new Map(); 
                m.set('a', 12);
                m.set('abc', -1161);
                m.set('b', 64);
            
                ({ 'tag': 'DeeperMap', 'map': m })
            ").unwrap()
        ).unwrap()
    );
```

Comment (awestlake87) [reply src/de.rs:239]: I uploaded 793f42b with the changes to tests and an update to the README. I think this should cover what we talked about before I brought up df09c8b, but let me know if you think we should add some more info or tests.

**Comment by attente** [comment]
Hi, is there any chance of this getting merged and released on crates.io? It would be great to have this fixed!

**Comment by egasimus** [comment]
Hi folks, what's the status on this?

**Review by RReverser — APPROVED** [review]
Ok, I've realised that as much as I wanted to try and do some final clean-ups here, this gets too tricky due to Serde's handling of certain types, and I won't find time to figure it out further.

The tests are looking good, and I trust your implementation, so I'm just doing to merge this and release as a new version on crates.io, and if there are more edge cases we could handle, we'll hopefully get further issues for them and address separately.

Sorry for blocking everyone for so long.

**Comment by RReverser** [comment]
Published as 0.2.0. Could probably be 0.1.x, but the complexity of some changes made me uncomfortable about potentially breaking existing code.
